src/langcheck/metrics/custom_text_quality.py

```diff
@@ -148,3 +154,232 @@ def _args_to_prompt_param(
         metric_values=scores,
         language=language,
     )
+
+
+def custom_pairwise_evaluator(
+    generated_outputs_a: list[str] | str | None,
+    generated_outputs_b: list[str] | str | None,
+    prompts: list[str] | str | None,
+    sources_a: list[str] | str | None,
+    sources_b: list[str] | str | None,
+    reference_outputs: list[str] | str | None,
+    eval_model: EvalClient,
+    metric_name: str,
+    score_map: dict[str, float],
+    template_path: str,
+    language: str,
+    enforce_consistency: bool = True,
+) -> MetricValue[float | None]:
+    """Calculates the scores of a custom pairwise evaluator, where "pairwise"
+    means that the Responses and/or Sources of two systems will be compared
+    against each other. The EvalClient will first assess the provided inputs
+    using the prompt template, and then convert those assessments into scores
+    using the score map.
+
+    The prompt template should be a Jinja2 file (file extension .j2) that
+    specifies the criteria that an LLM (as configured in the Eval Client) should
+    follow when evaluating an instance. The template is allowed to have
+    placeholders for the following variables (NOTE: not all are required):
+    - `gen_output_a`: Model A's generated output
+    - `gen_output_b`: Model B's generated output
+    - `user_query`: The prompt
+    - `src_a`: The source text for Model A
+    - `src_b`: The source text for Model B
+    - `ref_output`: The reference output
+
+    The prompt template should also specify the final available assessments for
+    the LLM evaluator, e.g. "Response A", "Response B", "Tie", etc. The score
+    map should then map each of those available assessments to a numerical
+    score. E.g. if the available assessments in the prompt template are
+    "Response A", "Response B", and "Tie", the score map should be something
+    like:
+    ``score_map = {'Response A': 0.0, 'Response B': 1.0, 'Tie': 0.5}``
```

Comment:
nit: it might be too obvious but we should mention that the options are symmetric between Response A and Response B, otherwise the consistency check will always fail.

Reply:
sorry for the dumb q, but what did you mean by symmetric here? or put another way, what would be an example of an asymmetric score map?

(Also noticed that I forgot to add the `enforce_consistency` docstring, so added that)

Reply:
Hmm it might be too obvious, but for example, if there are 5 options, 1 (for Response A) should correspond to 5 (for Response B), 2 should correspond to 4, 3 should be completely in the middle.

I don't think there is any reasonable scoring that doesn't follow this, but since if you don't follow that rule `enforce_consistency` always fails, I think it's worth noting.

Reply:
Ah ok I see what you mean. So if a user comes up with a scoring system like:
1 - Response A is way better
2 - Response A is slightly better
3 - Response B is slightly better

And Model A is always slightly better than Model B, then in the original order it'll always be 2 but then when swapped it'll always be 3.

I feel like it could be a bit distracting to put this in the top-level docstring though (I think the user would need to be intentionally unreasonable to come up with something like this 😆) so I'll add this as a code comment!

Reply:
added a quick note in the docstring, and a long explanation in the code comments